packages/strapi-plugin-users-permissions/config/functions/bootstrap.js

```diff
@@ -34,49 +34,52 @@ module.exports = async cb => {
     name: 'users-permissions'
   });
 
-  if (!await pluginStore.get({key: 'grant'})) {
-    const value = {
-      email: {
-        enabled: true,
-        icon: 'envelope'
-      },
-      facebook: {
-        enabled: false,
-        icon: 'facebook-official',
-        key: '',
-        secret: '',
-        callback: '/auth/facebook/callback',
-        scope: ['email']
-      },
-      google: {
-        enabled: false,
-        icon: 'google',
-        key: '',
-        secret: '',
-        callback: '/auth/google/callback',
-        scope: ['email']
-      },
-      github: {
-        enabled: false,
-        icon: 'github',
-        key: '',
-        secret: '',
-        redirect_uri: '/auth/github/callback',
-        scope: [
-          'user',
-          'user:email'
-        ]
-      },
-      twitter: {
-        enabled: false,
-        icon: 'twitter',
-        key: '',
-        secret: '',
-        callback: '/auth/twitter/callback'
-      }
-    };
-
-    await pluginStore.set({key: 'grant', value});
+  const grantValue = {
+    email: {
+      enabled: true,
+      icon: 'envelope'
+    },
+    facebook: {
+      enabled: false,
+      icon: 'facebook-official',
+      key: '',
+      secret: '',
+      callback: '/auth/facebook/callback',
+      scope: ['email']
+    },
+    google: {
+      enabled: false,
+      icon: 'google',
+      key: '',
+      secret: '',
+      callback: '/auth/google/callback',
+      scope: ['email']
+    },
+    github: {
+      enabled: false,
+      icon: 'github',
+      key: '',
+      secret: '',
+      redirect_uri: '/auth/github/callback',
+      scope: [
+        'user',
+        'user:email'
+      ]
+    },
+    twitter: {
+      enabled: false,
+      icon: 'twitter',
+      key: '',
+      secret: '',
+      callback: '/auth/twitter/callback'
+    }
+  };
+  const prevGrantValue = await pluginStore.get({key: 'grant'})
+  // store grant auth config to db
+  // when plugin_users-permissions_grant is not existed in db
+  // or we have added/deleted provider here.
+  if (!prevGrantValue || !_.isEqual(_.keys(prevGrantValue), _.keys(grantValue))) {
+    await pluginStore.set({key: 'grant', value: grantValue});
```

Comment:
So @zhuang17 can you just add merge of `prevGrantValue` and `grantValue` for the value to not loosepvrevious saved data if `prevGrantValue` is not empty.

Reply:
@lauriejim 

do you want rough merge like this,  
```js
const newGrantValue = _.merge (grantValue, prevGrantValue)  // or {...grantValue, ...prevGrantValue}
```

or want a more precise merge ( I mean only merge previous values for those property which still exist in new config now).
since the `{...grantValue, ...prevGrantValue}`  way will merge some extra config even if we have delete the property in new cofig ?

maybe something like this ?
```js
Object.keys(grantValue).forEach((key) => {
    if (key in prevGrantValue) { // or obj1.hasOwnProperty(key)
        grantValue[key] = prevGrantValue[key];
    }
});

// and may do it Recursively  for nested config object
```

Reply:
`_.merge` will do the job for what you want.